### ISSUE TITLE
[Bug] PointPillars trained with KITTI on different classes ('truck')  is not able to learn

### ISSUE THREAD

**Issue by OttaviaBelotti** [open-mmlab/mmdetection3d#2843]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmdetection3d/issues) and [Discussions](https://github.com/open-mmlab/mmdetection3d/discussions) but cannot get the expected help.
- [X] I have read the [FAQ documentation](https://mmdetection3d.readthedocs.io/en/latest/notes/faq.html) but cannot get the expected help.
- [X] The bug has not been fixed in the [latest version (dev-1.x)](https://github.com/open-mmlab/mmdetection3d/tree/dev-1.x) or [latest version (dev-1.0)](https://github.com/open-mmlab/mmdetection3d/tree/dev-1.0).

### Task

I have modified the scripts/configs, or I'm working on my own tasks/models/datasets.

### Branch

1.x branch https://github.com/open-mmlab/mmdetection3d/tree/dev-1.x

### Environment

sys.platform: linux
Python: 3.8.10 (default, May 26 2023, 14:05:08) [GCC 9.4.0]
CUDA available: True
numpy_random_seed: 2147483648
GPU 0: NVIDIA GeForce RTX 3090
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 11.3, V11.3.109
GCC: x86_64-linux-gnu-gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
PyTorch: 1.12.1+cu113
PyTorch compiling details: PyTorch built with:
  - GCC 9.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86
  - CuDNN 8.3.2  (built against CUDA 11.5)
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.3.2, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -fabi-version=11 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.12.1, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, 

TorchVision: 0.13.1+cu113
OpenCV: 4.8.1
MMEngine: 0.10.1
MMDetection: 3.2.0
MMDetection3D: 1.2.0+63d8937
spconv2.0: False

### Reproduces the problem - code sample

configs/pointpillars/pointpillars_hv_secfpn_8xb6-160e_kitti-3d-3class_tanh.py
```python
_base_ = [
    '../_base_/models/pointpillars_hv_secfpn_kitti_full_class.py',
    '../_base_/datasets/kitti-3d-full_class.py',
    '../_base_/schedules/cyclic-40e.py', '../_base_/default_runtime.py'
]

point_cloud_range = [0, -39.68, -3, 69.12, 39.68, 1]
# dataset settings
data_root = '/DATA/datasets/kitti_tanh/'
class_names = ['Pedestrian', 'Cyclist', 'Car', 'Truck']
metainfo = dict(classes=class_names)
backend_args = None

# PointPillars adopted a different sampling strategies among classes
db_sampler = dict(
    data_root=data_root,
    info_path=data_root + 'kitti_dbinfos_train.pkl',
    rate=1.0,
    prepare=dict(
        filter_by_difficulty=[-1],
        filter_by_min_points=dict(Car=5, Pedestrian=5, Cyclist=5, Truck=5)),
    classes=class_names,
    sample_groups=dict(Car=15, Pedestrian=15, Cyclist=15, Truck=15),
    points_loader=dict(
        type='LoadPointsFromFile',
        coord_type='LIDAR',
        load_dim=4,
        use_dim=4,
        backend_args=backend_args),
    backend_args=backend_args)

# PointPillars uses different augmentation hyper parameters
train_pipeline = [
    dict(
        type='LoadPointsFromFile',
        coord_type='LIDAR',
        load_dim=4,
        use_dim=4,
        backend_args=backend_args),
    dict(type='LoadAnnotations3D', with_bbox_3d=True, with_label_3d=True),
    dict(type='ObjectSample', db_sampler=db_sampler, use_ground_plane=False),
    dict(type='RandomFlip3D', flip_ratio_bev_horizontal=0.5),
    dict(
        type='GlobalRotScaleTrans',
        rot_range=[-0.78539816, 0.78539816],
        scale_ratio_range=[0.95, 1.05]),
    dict(type='PointsRangeFilter', point_cloud_range=point_cloud_range),
    dict(type='ObjectRangeFilter', point_cloud_range=point_cloud_range),
    dict(type='PointShuffle'),
    dict(
        type='Pack3DDetInputs',
        keys=['points', 'gt_labels_3d', 'gt_bboxes_3d'])
]
test_pipeline = [
    dict(
        type='LoadPointsFromFile',
        coord_type='LIDAR',
        load_dim=4,
        use_dim=4,
        backend_args=backend_args),
    dict(
        type='MultiScaleFlipAug3D',
        img_scale=(1333, 800),
        pts_scale_ratio=1,
        flip=False,
        transforms=[
            dict(
                type='GlobalRotScaleTrans',
                rot_range=[0, 0],
                scale_ratio_range=[1., 1.],
                translation_std=[0, 0, 0]),
            dict(type='RandomFlip3D'),
            dict(
                type='PointsRangeFilter', point_cloud_range=point_cloud_range)
        ]),
    dict(type='Pack3DDetInputs', keys=['points'])
]

train_dataloader = dict(
    dataset=dict(dataset=dict(pipeline=train_pipeline, metainfo=metainfo)))
test_dataloader = dict(dataset=dict(pipeline=test_pipeline, metainfo=metainfo))
val_dataloader = dict(dataset=dict(pipeline=test_pipeline, metainfo=metainfo))
# In practice PointPillars also uses a different schedule
# optimizer
lr = 0.001
epoch_num = 80
optim_wrapper = dict(
     optimizer=dict(lr=lr), clip_grad=dict(max_norm=35, norm_type=2))
param_scheduler = [
    dict(
        type='CosineAnnealingLR',
        T_max=epoch_num * 0.4,
        eta_min=lr * 10,
        begin=0,
        end=epoch_num * 0.4,
        by_epoch=True,
        convert_to_iter_based=True),
    dict(
        type='CosineAnnealingLR',
        T_max=epoch_num * 0.6,
        eta_min=lr * 1e-4,
        begin=epoch_num * 0.4,
        end=epoch_num * 1,
        by_epoch=True,
        convert_to_iter_based=True),
    dict(
        type='CosineAnnealingMomentum',
        T_max=epoch_num * 0.4,
        eta_min=0.85 / 0.95,
        begin=0,
        end=epoch_num * 0.4,
        by_epoch=True,
        convert_to_iter_based=True),
    dict(
        type='CosineAnnealingMomentum',
        T_max=epoch_num * 0.6,
        eta_min=1,
        begin=epoch_num * 0.4,
        end=epoch_num * 1,
        convert_to_iter_based=True)
]
# max_norm=35 is slightly better than 10 for PointPillars in the earlier
# development of the codebase thus we keep the setting. But we does not
# specifically tune this parameter.
# PointPillars usually need longer schedule than second, we simply double
# the training schedule. Do remind that since we use RepeatDataset and
# repeat factor is 2, so we actually train 160 epochs.
train_cfg = dict(by_epoch=True, max_epochs=epoch_num, val_interval=2)
val_cfg = dict()
test_cfg = dict()
```
-----------------------------------------------------------------------
configs/\_base\_/datasets/kitti-3d-full_class.py
```python
# dataset settings
dataset_type = 'KittiDataset'
data_root = '/DATA/datasets/kitti_tanh/'
class_names = ['Pedestrian', 'Cyclist', 'Car', 'Truck']
point_cloud_range = [0, -40, -3, 70.4, 40, 1]
input_modality = dict(use_lidar=True, use_camera=False)
metainfo = dict(classes=class_names)

# Example to use different file client
# Method 1: simply set the data root and let the file I/O module
# automatically infer from prefix (not support LMDB and Memcache yet)

# data_root = 's3://openmmlab/datasets/detection3d/kitti/'

# Method 2: Use backend_args, file_client_args in versions before 1.1.0
# backend_args = dict(
#     backend='petrel',
#     path_mapping=dict({
#         './data/': 's3://openmmlab/datasets/detection3d/',
#          'data/': 's3://openmmlab/datasets/detection3d/'
#      }))
backend_args = None

db_sampler = dict(
    data_root=data_root,
    info_path=data_root + 'kitti_dbinfos_train.pkl',
    rate=1.0,
    prepare=dict(
        filter_by_difficulty=[-1],
        filter_by_min_points=dict(Car=5, Pedestrian=10, Cyclist=10, Truck=5)),
    classes=class_names,
    sample_groups=dict(Car=12, Pedestrian=6, Cyclist=6, Truck=6),
    points_loader=dict(
        type='LoadPointsFromFile',
        coord_type='LIDAR',
        load_dim=4,
        use_dim=4,
        backend_args=backend_args),
    backend_args=backend_args)

train_pipeline = [
    dict(
        type='LoadPointsFromFile',
        coord_type='LIDAR',
        load_dim=4,  # x, y, z, intensity
        use_dim=4,
        backend_args=backend_args),
    dict(type='LoadAnnotations3D', with_bbox_3d=True, with_label_3d=True),
    dict(type='ObjectSample', db_sampler=db_sampler),
    dict(
        type='ObjectNoise',
        num_try=100,
        translation_std=[1.0, 1.0, 0.5],
        global_rot_range=[0.0, 0.0],
        rot_range=[-0.78539816, 0.78539816]),
    dict(type='RandomFlip3D', flip_ratio_bev_horizontal=0.5),
    dict(
        type='GlobalRotScaleTrans',
        rot_range=[-0.78539816, 0.78539816],
        scale_ratio_range=[0.95, 1.05]),
    dict(type='PointsRangeFilter', point_cloud_range=point_cloud_range),
    dict(type='ObjectRangeFilter', point_cloud_range=point_cloud_range),
    dict(type='PointShuffle'),
    dict(
        type='Pack3DDetInputs',
        keys=['points', 'gt_bboxes_3d', 'gt_labels_3d'])
]
test_pipeline = [
    dict(
        type='LoadPointsFromFile',
        coord_type='LIDAR',
        load_dim=4,
        use_dim=4,
        backend_args=backend_args),
    dict(
        type='MultiScaleFlipAug3D',
        img_scale=(1333, 800),
        pts_scale_ratio=1,
        flip=False,
        transforms=[
            dict(
                type='GlobalRotScaleTrans',
                rot_range=[0, 0],
                scale_ratio_range=[1., 1.],
                translation_std=[0, 0, 0]),
            dict(type='RandomFlip3D'),
            dict(
                type='PointsRangeFilter', point_cloud_range=point_cloud_range)
        ]),
    dict(type='Pack3DDetInputs', keys=['points'])
]
# construct a pipeline for data and gt loading in show function
# please keep its loading function consistent with test_pipeline (e.g. client)
eval_pipeline = [
    dict(
        type='LoadPointsFromFile',
        coord_type='LIDAR',
        load_dim=4,
        use_dim=4,
        backend_args=backend_args),
    dict(type='Pack3DDetInputs', keys=['points'])
]
train_dataloader = dict(
    batch_size=6,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    dataset=dict(
        type='RepeatDataset',
        times=2,
        dataset=dict(
            type=dataset_type,
            data_root=data_root,
            ann_file='kitti_infos_train.pkl',
            data_prefix=dict(pts='training/velodyne_reduced'),
            pipeline=train_pipeline,
            modality=input_modality,
            test_mode=False,
            metainfo=metainfo,
            # we use box_type_3d='LiDAR' in kitti and nuscenes dataset
            # and box_type_3d='Depth' in sunrgbd and scannet dataset.
            box_type_3d='LiDAR',
            backend_args=backend_args)))
val_dataloader = dict(
    batch_size=1,
    num_workers=1,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        data_prefix=dict(pts='training/velodyne_reduced'),
        ann_file='kitti_infos_val.pkl',
        pipeline=test_pipeline,
        modality=input_modality,
        test_mode=True,
        metainfo=metainfo,
        box_type_3d='LiDAR',
        backend_args=backend_args))
test_dataloader = dict(
    batch_size=1,
    num_workers=1,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        data_prefix=dict(pts='training/velodyne_reduced'),
        ann_file='kitti_infos_val.pkl',
        pipeline=test_pipeline,
        modality=input_modality,
        test_mode=True,
        metainfo=metainfo,
        box_type_3d='LiDAR',
        backend_args=backend_args))
val_evaluator = dict(
    type='KittiMetric',
    ann_file=data_root + 'kitti_infos_val.pkl',
    metric='bbox',
    backend_args=backend_args)
test_evaluator = val_evaluator

vis_backends = [dict(type='LocalVisBackend')]
visualizer = dict(
    type='Det3DLocalVisualizer', vis_backends=vis_backends, name='visualizer')

```
-----------------------------------------------------------------------
configs/\_base\_/models/pointpillars_hv_secfpn_kitti_full_class.py
```python
voxel_size = [0.16, 0.16, 4]

model = dict(
    type='VoxelNet',
    data_preprocessor=dict(
        type='Det3DDataPreprocessor',
        voxel=True,
        voxel_layer=dict(
            max_num_points=32,  # max_points_per_voxel
            point_cloud_range=[0, -39.68, -3, 69.12, 39.68, 1],
            voxel_size=voxel_size,
            max_voxels=(16000, 40000))),
    voxel_encoder=dict(
        type='PillarFeatureNet',
        in_channels=4,
        feat_channels=[64],
        with_distance=False,
        voxel_size=voxel_size,
        point_cloud_range=[0, -39.68, -3, 69.12, 39.68, 1]),
    middle_encoder=dict(
        type='PointPillarsScatter', in_channels=64, output_shape=[496, 432]),
    backbone=dict(
        type='SECOND',
        in_channels=64,
        layer_nums=[3, 5, 5],
        layer_strides=[2, 2, 2],
        out_channels=[64, 128, 256]),
    neck=dict(
        type='SECONDFPN',
        in_channels=[64, 128, 256],
        upsample_strides=[1, 2, 4],
        out_channels=[128, 128, 128]),
    bbox_head=dict(
        type='Anchor3DHead',
        num_classes=4,
        in_channels=384,
        feat_channels=384,
        use_direction_classifier=True,
        assign_per_class=True,
        anchor_generator=dict(
            type='AlignedAnchor3DRangeGenerator',
            ranges=[
                [0, -39.68, -0.6, 69.12, 39.68, -0.6],
                [0, -39.68, -0.6, 69.12, 39.68, -0.6],
                [0, -39.68, -1.78, 69.12, 39.68, -1.78],
                [0, -39.68, -1.78, 69.12, 39.68, -1.78],
            ],
            sizes=[[0.8, 0.6, 1.73], [1.76, 0.6, 1.73], [3.9, 1.6, 1.56], [10.729741379310346, 2.6485919540229883, 3.0977011494252875]],
            rotations=[0, 1.57],
            reshape_out=False),
        diff_rad_by_sin=True,
        bbox_coder=dict(type='DeltaXYZWLHRBBoxCoder'),
        loss_cls=dict(
            type='mmdet.FocalLoss',
            use_sigmoid=True,
            gamma=2.0,
            alpha=0.25,
            loss_weight=1.0),
        loss_bbox=dict(
            type='mmdet.SmoothL1Loss', beta=1.0 / 9.0, loss_weight=2.0),
        loss_dir=dict(
            type='mmdet.CrossEntropyLoss', use_sigmoid=False,
            loss_weight=0.2)),
    # model training and testing settings
    train_cfg=dict(
        assigner=[
            dict(  # for Pedestrian
                type='Max3DIoUAssigner',
                iou_calculator=dict(type='mmdet3d.BboxOverlapsNearest3D'),
                pos_iou_thr=0.5,
                neg_iou_thr=0.35,
                min_pos_iou=0.35,
                ignore_iof_thr=-1),
            dict(  # for Cyclist
                type='Max3DIoUAssigner',
                iou_calculator=dict(type='mmdet3d.BboxOverlapsNearest3D'),
                pos_iou_thr=0.5,
                neg_iou_thr=0.35,
                min_pos_iou=0.35,
                ignore_iof_thr=-1),
            dict(  # for Car
                type='Max3DIoUAssigner',
                iou_calculator=dict(type='mmdet3d.BboxOverlapsNearest3D'),
                pos_iou_thr=0.6,
                neg_iou_thr=0.45,
                min_pos_iou=0.45,
                ignore_iof_thr=-1),
            dict(  # for Truck
                type='Max3DIoUAssigner',
                iou_calculator=dict(type='mmdet3d.BboxOverlapsNearest3D'),
                pos_iou_thr=0.5,
                neg_iou_thr=0.35,
                min_pos_iou=0.35,
                ignore_iof_thr=-1)
        ],
        allowed_border=0,
        pos_weight=-1,
        debug=False),
    test_cfg=dict(
        use_rotate_nms=True,
        nms_across_levels=False,
        nms_thr=0.01,
        score_thr=0.1,
        min_bbox_size=0,
        nms_pre=100,
        max_num=50))

```

### Reproduces the problem - command or script

```shelll
python3 tools/train.py configs/pointpillars/pointpillars_hv_secfpn_8xb6-160e_kitti-3d-3class_tanh.py
```

### Reproduces the problem - error message

None

### Additional information

In this case I'm trying to train PointPillars on a modified version of KITTI dataset, where the modfication implies a remapping of every point's intensity value in the point cloud samples applying a tanh estimator. The final intensity distribution for each pointcloud has always mean=0.5, std=0.005 and in the range [0.49, 0.525].
The issue arises when I switch from a 3-classes evaluation benchmark ('pedestrian', 'cyclist', 'car') to a 4-classes one ( 'pedestrian', 'cyclist', 'car', 'truck'), both in this case (with the intensity values remapped) and with the regular KITTI one. In the 3-classes case I can achieve decent results (here I report the results for the remapped intensity version):
```log
----------- AP11 Results ------------

Pedestrian AP11@0.50, 0.50, 0.50:
bbox AP11:59.8557, 56.4460, 52.9005
bev  AP11:62.4412, 55.9147, 52.5105
3d   AP11:54.2235, 49.5573, 45.3896
aos  AP11:27.04, 26.67, 24.69
Pedestrian AP11@0.50, 0.25, 0.25:
bbox AP11:59.8557, 56.4460, 52.9005
bev  AP11:72.6835, 68.6748, 64.0063
3d   AP11:72.5300, 68.2527, 63.5352
aos  AP11:27.04, 26.67, 24.69
Cyclist AP11@0.50, 0.50, 0.50:
bbox AP11:63.5019, 49.8066, 48.1614
bev  AP11:61.4137, 45.4912, 42.7067
3d   AP11:59.3237, 42.0993, 40.1008
aos  AP11:47.39, 35.90, 34.85
Cyclist AP11@0.50, 0.25, 0.25:
bbox AP11:63.5019, 49.8066, 48.1614
bev  AP11:65.5098, 51.0992, 47.8436
3d   AP11:65.5098, 51.0992, 47.8436
aos  AP11:47.39, 35.90, 34.85
Car AP11@0.70, 0.70, 0.70:
bbox AP11:90.2899, 88.5813, 86.2440
bev  AP11:89.2551, 84.2702, 79.2077
3d   AP11:83.3963, 73.5696, 67.9455
aos  AP11:89.84, 87.71, 84.93
Car AP11@0.70, 0.50, 0.50:
bbox AP11:90.2899, 88.5813, 86.2440
bev  AP11:90.5116, 89.5048, 89.0815
3d   AP11:90.5084, 89.4091, 88.8300
aos  AP11:89.84, 87.71, 84.93

Overall AP11@easy, moderate, hard:
bbox AP11:71.2159, 64.9446, 62.4353
bev  AP11:71.0367, 61.8920, 58.1416
3d   AP11:65.6479, 55.0754, 51.1453
aos  AP11:54.76, 50.09, 48.16

----------- AP40 Results ------------

Pedestrian AP40@0.50, 0.50, 0.50:
bbox AP40:59.1424, 55.1348, 51.4939
bev  AP40:61.9177, 55.4458, 51.0751
3d   AP40:53.6765, 47.6234, 43.5336
aos  AP40:27.00, 26.33, 24.58
Pedestrian AP40@0.50, 0.25, 0.25:
bbox AP40:59.1424, 55.1348, 51.4939
bev  AP40:73.8091, 68.8170, 64.0752
3d   AP40:73.6248, 68.2859, 63.5430
aos  AP40:27.00, 26.33, 24.58
Cyclist AP40@0.50, 0.50, 0.50:
bbox AP40:64.5232, 49.4491, 46.3212
bev  AP40:61.5885, 43.5248, 41.1278
3d   AP40:58.8859, 40.2027, 37.5920
aos  AP40:45.98, 32.94, 30.85
Cyclist AP40@0.50, 0.25, 0.25:
bbox AP40:64.5232, 49.4491, 46.3212
bev  AP40:65.1457, 49.6273, 46.8747
3d   AP40:65.1457, 49.6273, 46.8747
aos  AP40:45.98, 32.94, 30.85
Car AP40@0.70, 0.70, 0.70:
bbox AP40:95.5214, 89.5606, 87.1554
bev  AP40:92.2193, 85.4992, 83.2586
3d   AP40:84.2660, 73.0955, 70.6861
aos  AP40:94.99, 88.64, 85.77
Car AP40@0.70, 0.50, 0.50:
bbox AP40:95.5214, 89.5606, 87.1554
bev  AP40:95.6177, 94.4331, 92.1300
3d   AP40:95.8994, 92.3587, 89.9869
aos  AP40:94.99, 88.64, 85.77

Overall AP40@easy, moderate, hard:
bbox AP40:73.0623, 64.7148, 61.6568
bev  AP40:71.9085, 61.4899, 58.4872
3d   AP40:65.6095, 53.6405, 50.6039
aos  AP40:55.99, 49.30, 47.07
```
while when I switch to the 4-classes case, the net is not able to learn at all (even for the car, predestrian and cyclist classes):
```log
----------- AP11 Results ------------

Pedestrian AP11@0.50, 0.50, 0.50:
bbox AP11:0.1196, 4.5455, 4.5455
bev  AP11:0.0000, 0.0000, 0.0000
3d   AP11:0.0000, 0.0000, 0.0000
aos  AP11:0.12, 0.16, 0.16
Pedestrian AP11@0.50, 0.25, 0.25:
bbox AP11:0.1196, 4.5455, 4.5455
bev  AP11:0.0035, 0.0059, 0.0076
3d   AP11:0.0026, 0.0045, 0.0059
aos  AP11:0.12, 0.16, 0.16
Cyclist AP11@0.50, 0.50, 0.50:
bbox AP11:0.0063, 0.0060, 0.0060
bev  AP11:0.0000, 0.0000, 0.0000
3d   AP11:0.0000, 0.0000, 0.0000
aos  AP11:0.00, 0.00, 0.00
Cyclist AP11@0.50, 0.25, 0.25:
bbox AP11:0.0063, 0.0060, 0.0060
bev  AP11:0.0000, 0.0000, 0.0000
3d   AP11:0.0000, 0.0000, 0.0000
aos  AP11:0.00, 0.00, 0.00
Car AP11@0.70, 0.70, 0.70:
bbox AP11:1.5152, 1.5152, 1.5152
bev  AP11:0.0694, 0.0649, 0.0649
3d   AP11:0.0021, 0.0059, 0.0063
aos  AP11:1.52, 1.52, 1.52
Car AP11@0.70, 0.50, 0.50:
bbox AP11:1.5152, 1.5152, 1.5152
bev  AP11:0.7576, 0.7576, 0.7576
3d   AP11:0.7576, 0.7576, 0.7576
aos  AP11:1.52, 1.52, 1.52
Truck AP11@0.70, 0.70, 0.70:
bbox AP11:0.0000, 0.0000, 0.0000
bev  AP11:0.0000, 0.0000, 0.0000
3d   AP11:0.0000, 0.0000, 0.0000
aos  AP11:0.00, 0.00, 0.00
Truck AP11@0.70, 0.50, 0.50:
bbox AP11:0.0000, 0.0000, 0.0000
bev  AP11:0.0000, 0.0062, 0.0062
3d   AP11:0.0000, 0.0000, 0.0000
aos  AP11:0.00, 0.00, 0.00

Overall AP11@easy, moderate, hard:
bbox AP11:0.4103, 1.5166, 1.5166
bev  AP11:0.0173, 0.0162, 0.0162
3d   AP11:0.0005, 0.0015, 0.0016
aos  AP11:0.41, 0.42, 0.42

----------- AP40 Results ------------

Pedestrian AP40@0.50, 0.50, 0.50:
bbox AP40:0.0042, 0.0059, 0.0069
bev  AP40:0.0000, 0.0000, 0.0000
3d   AP40:0.0000, 0.0000, 0.0000
aos  AP40:0.00, 0.00, 0.00
Pedestrian AP40@0.50, 0.25, 0.25:
bbox AP40:0.0042, 0.0059, 0.0069
bev  AP40:0.0010, 0.0016, 0.0021
3d   AP40:0.0007, 0.0012, 0.0016
aos  AP40:0.00, 0.00, 0.00
Cyclist AP40@0.50, 0.50, 0.50:
bbox AP40:0.0013, 0.0016, 0.0016
bev  AP40:0.0000, 0.0000, 0.0000
3d   AP40:0.0000, 0.0000, 0.0000
aos  AP40:0.00, 0.00, 0.00
Cyclist AP40@0.50, 0.25, 0.25:
bbox AP40:0.0013, 0.0016, 0.0016
bev  AP40:0.0000, 0.0000, 0.0000
3d   AP40:0.0000, 0.0000, 0.0000
aos  AP40:0.00, 0.00, 0.00
Car AP40@0.70, 0.70, 0.70:
bbox AP40:0.0375, 0.0574, 0.0680
bev  AP40:0.0027, 0.0052, 0.0058
3d   AP40:0.0006, 0.0016, 0.0017
aos  AP40:0.03, 0.04, 0.05
Car AP40@0.70, 0.50, 0.50:
bbox AP40:0.0375, 0.0574, 0.0680
bev  AP40:0.0339, 0.0487, 0.0271
3d   AP40:0.0219, 0.0160, 0.0179
aos  AP40:0.03, 0.04, 0.05
Truck AP40@0.70, 0.70, 0.70:
bbox AP40:0.0000, 0.0000, 0.0000
bev  AP40:0.0000, 0.0000, 0.0000
3d   AP40:0.0000, 0.0000, 0.0000
aos  AP40:0.00, 0.00, 0.00
Truck AP40@0.70, 0.50, 0.50:
bbox AP40:0.0000, 0.0000, 0.0000
bev  AP40:0.0000, 0.0000, 0.0000
3d   AP40:0.0000, 0.0000, 0.0000
aos  AP40:0.00, 0.00, 0.00

Overall AP40@easy, moderate, hard:
bbox AP40:0.0108, 0.0162, 0.0191
bev  AP40:0.0007, 0.0013, 0.0014
3d   AP40:0.0001, 0.0004, 0.0004
aos  AP40:0.01, 0.01, 0.01
```

I have regenerated the information and gt_database to be sure even if it might not be necessary: `python3 tools/create_data.py kitti --root-path /DATA/datasets/kitti_tanh --out-dir /DATA/datasets/kitti_tanh --extra-tag kitti`.
I also had to modify `mmdet3d/datasets/kitti_dataset.py` and `mmdet3d/evaluation/functional/kitti_utils/eval.py` because there were only 5 hardcoded classes while 'truck' was missing:
```python
def clean_data(gt_anno, dt_anno, current_class, difficulty):
    CLASS_NAMES = ['car', 'pedestrian', 'cyclist', 'person_sitting', 'van', 'truck', 'tram']
    MIN_HEIGHT = [40, 25, 25]
    MAX_OCCLUSION = [0, 1, 2]
    MAX_TRUNCATION = [0.15, 0.3, 0.5]
...

# @ def kitti_eval(...)
    Args:
    assert len(eval_types) > 0, 'must contain at least one evaluation type'
    if 'aos' in eval_types:
        assert 'bbox' in eval_types, 'must evaluate bbox when evaluating aos'
    overlap_0_7 = np.array([[0.7, 0.5, 0.5, 0.7, 0.5, 0.7, 0.7], 
                            [0.7, 0.5, 0.5, 0.7, 0.5, 0.7, 0.7],
                            [0.7, 0.5, 0.5, 0.7, 0.5, 0.7, 0.7]])
    overlap_0_5 = np.array([[0.7, 0.5, 0.5, 0.7, 0.5, 0.7, 0.7],
                            [0.5, 0.25, 0.25, 0.5, 0.25, 0.5, 0.5],
                            [0.5, 0.25, 0.25, 0.5, 0.25, 0.5, 0.5]])
    min_overlaps = np.stack([overlap_0_7, overlap_0_5], axis=0)  # [2, 3, 5]
    class_to_name = {
        0: 'Car',
        2: 'Cyclist',
        3: 'Van',
        4: 'Person_sitting',
        5: 'Truck',
        6: 'Tram'
    }
    name_to_class = {v: n for n, v in class_to_name.items()}
    if not isinstance(current_classes, (list, tuple)):
...
```

In conclusion, have I done something wrong with the config files to switch from 3 classes to 4 classes or is it something deeper? What would you suggest to do to actually allow the net to learn even with the 'truck' class?
